### PR TITLE
WIP: add SQUAREM for fixed point problems

### PR DESCRIPTION
**Pull request by pv**
This adds the SQUAREM algorithm plus a root() like interface for fixed point problems.

There's probably stuff to discuss on the interface side, but also on the algorithm side: The backtracking is as suggested in the paper, but I'm not so sure if that's a very good choice in general, as it probably requires the mapping is a contraction. Also, someone who actually does EM would probably need to comment whether this would be useful for them or not.

I think there's room for adding more fixed point solvers --- although the problem can always be mapped to root finding and vice versa, they're not exactly the same thing. Besides, the steffensen/aitken method offered by Scipy currently looks a bit naive...

### CODE REVIEW

**Comment by argriffing** [comment]
> The backtracking is as suggested in the paper, but I'm not so sure if that's a very good choice in general, as it probably requires the mapping is a contraction. Also, someone who actually does EM would probably need to comment whether this would be useful for them or not.

I've used EM, and acceleration algorithms like SQUAREM that use EM's extra structure would certainly be useful, although as you've mentioned they may not be appropriate as general fixed-point or root-finding solvers.  If this is added to scipy (which I hope it will be) then it would be helpful to clarify its scope, for example by defining "monotone" and "contraction mapping" as used in the description
`Globally-convergent, partially monotone, acceleration schemes for accelerating the convergence
of any smooth, monotone, slowly-converging contraction mapping.`  I think this means you have two functions `func` and `fmerit` so that the sequence
x0 = x0
x1 = func(x0)
x2 = func(func(x0))
x3 = func(func(func(x0)))
...
is monotonically decreasing in the merit function, in the sense that `fmerit(x_{i+1}) < fmerit(x_i)` even if `fmerit` is not explicitly provided, and that the [`contraction mapping`](http://en.wikipedia.org/wiki/Contraction_mapping) is a property of `func` which somehow implies that the sequence `x_i` approaches a unique fixed point.  I'm not completely sure that what I've written above is correct, but I know that some subtle issues are involved.

**Comment by argriffing** [comment]
In trying to understand the merit function, I found the docstring
```
fmerit : callable
        Merit function fmerit(f) for computing merit value from residual.
```
which seems to have become outdated.  I think `fmerit` now takes two arguments, but I still don't understand exactly what it means.  I don't understand how it should be interpreted in the context of EM.  Maybe I just need to read the docs more thoroughly.

File: scipy/optimize/_spectral.py [hunk 1]
```diff
@@ -173,6 +191,150 @@ def fmerit(F):
...
+
+        (1)  alpha_k = vdot(r, v) / vdot(v, v)
+        (2)  alpha_k = vdot(r, r) / vdot(r, v)
+        (3)  alpha_k = -sqrt(vdot(r, r) / vdot(v, v))
+
+    The different choices minimize different step length model objective functions.
+
+    References
+    ----------
+    .. [1] R. Varadhan, C. Roland. Scand. J. Statistics, 35, 335 (2008).
```
Comment (josef-pkt) [inline scipy/optimize/_spectral.py]: add title ? easier to search for, although google found also by author and journal.

File: scipy/optimize/__init__.py [hunk 2]
```diff
@@ -110,7 +110,8 @@
 .. autosummary::
    :toctree: generated/
 
-   fixed_point - Single-variable fixed-point solver
+   fixpoint - Single-variable fixed-point solver
+   fixed_point - Steffensen's method for fixed points
```
Comment (josef-pkt) [inline scipy/optimize/__init__.py]: names sound a bit confusing, difference looks to me more like a typo than anything that would give a hint for the actual difference.
(I don't have any alternative suggestion.)

**Comment by pv** [comment]
In EM it is IIRC related to (negative?) likelihood.

File: scipy/optimize/__init__.py (hunk 2 above)
Comment (pv) [inline scipy/optimize/__init__.py]: The problem here is that the signature of fixed_point was fairly
specific for the steffensen/aitken naive iteration, and changing it
breaks backward compat.
.
The actual difference as it is in this PR is that fixed_point is deprecated.

**Comment by argriffing** [comment]
I'll plan to look at this in more detail.  I've actually implemented this before, but I never got around to submitting it to scipy/statsmodels because I've had problems reproducing published results on test cases and because EM acceleration expects more structure than generic fixed-point acceleration so it is less widely applicable.

**Comment by josef-pkt** [comment]
@argriffing  If you have time and interest, you could code the Poisson example from the article or another MLE example that is solved using EM.
It would make a good test case, and I would take it and blow it up to full model size for statsmodels.

(and it would be possible to compare with basin_hopping as an alternative optimizer, which I used in the past for mixture problems.)

File: scipy/optimize/__init__.py (hunk 2 above)
Comment (josef-pkt) [inline scipy/optimize/__init__.py]: fine, if fixed_point will eventually go away, then there won't be any confusion anymore.

**Comment by argriffing** [comment]
> If you have time and interest, you could code the Poisson example from the article

@josef-pkt I've tried this before, but I didn't seem to get results matching those in the paper.  Also, they report a summary of random starting points, so Table 3 is not be completely reproducible even with a correct implementation.  Maybe I'll try this again with @pv's implementation.

**Comment by josef-pkt** [comment]
> I've tried this before, but I didn't seem to get results matching those in the paper.

@argriffing The optimization "should" work, even if it doesn't match the article completely or very closely. They report 3% convergence failures in the Poisson mixture case, so there are some problems, and maybe they used some other tricks that they don't mention.
If there are convergence problems, then I would start with a smaller range of initial values, and look for numerical problems.
If the number of function evaluations is much larger in this PR than in the article, then it would be a good test case and show that there is room for improvement.

(A few years ago I coded an alternative version of the poisson mixture estimation, that uses standard gradient based methods but puts a Logit transformation on the mixture probability.)

**Comment by josef-pkt** [comment]
Besides, independently of how well it works, it would be a good example to illustrate the use of the solver in this PR which is advertised as general purpose EM accelerator. Once we know how to use it, we will be able to build real models following the same pattern.

**Comment by argriffing** [comment]
@josef-pkt Yes I'm working on a PR now.  FWIW my point in the comments above was not that the optimization worked badly!  I remember that it had worked very well.  My memory of the details is so vague that perhaps I was even getting some unreasonably better results than were reported in the paper, or not.

**Comment by argriffing** [comment]
I have the branch https://github.com/argriffing/scipy/tree/pv_squarem with a new test, and I'll try to remember how to send a PR into this PR.

**Comment by argriffing** [comment]
@pv I've sent https://github.com/pv/scipy-work/pull/10 which I think would go into this branch?

**Comment by argriffing** [comment]
> (A few years ago I coded an alternative version of the poisson mixture estimation, that uses standard gradient based methods but puts a Logit transformation on the mixture probability.)

This would be helpful if you have analytically derived gradients.  For another Poisson mixture example that I'm testing (example 1 of "Acceleration of the EM Algorithm by Using Quasi-Newton Methods") the generic L-BFGS-B is having trouble converging, maybe because its finite differences approximation of the gradient is too inexact?  I guess automatic differentiation is currently unavailable for scipy tests.

**Comment by argriffing** [comment]
The docstring says the SQUAREM default step rule is 2 but the code says 3.

**Comment by josef-pkt** [comment]
@argriffing about Poisson mixture: I don't remember the details, and I tried several versions including poisson with normal mixing probabilities. This was for panel data Poisson regression with a random intercept, and I was mostly just playing with experimental versions to see how it could work and be designed. I didn't use any analytical derivatives, IIRC.
A bit more recent and more serious, mixtures of VonMises distributions with multinomial probabilities for the mixing probabilities. We used mostly basin-hopping with bfgs, the results "looked" good (good match between estimated pdf and histogram), but the hessian had quite often some negative eigenvalues. The latter happened more often in overparameterized versions where one component was either small or might have been redundant.
https://github.com/rc/dist_mixtures/blob/master/dist_mixtures/mixture_von_mises.py#L232
IIRC, we ended up using mostly the Binned version, because the data was already in histogram form.
bfgs often ended up in local optima, fmin (Nelder-Mead) even got stuck in near optima where bfgs would not get stuck. basin-hopping worked very well.

**Comment by argriffing** [comment]
I agree that this needs tests that evaluate the log likelihood to steer the accelerated EM away from overly ambitious step lengths that decrease the log likelihood (this is always possible because the vanilla EM step will never decrease log likelihood).  I'll try to add something today.

Also I noticed that my bijection between a simplex and R^n is ad-hoc and has some bad properties, and that I should probably use the "isometric log ratio transform" instead.  Maybe this would be of independent interest in scipy?  I see a related old message on the mailing list when I searched 'scipy' and 'aitchison' http://mail.scipy.org/pipermail/scipy-user/2008-April/016629.html.

Another thing that is weird and ad-hoc about my sub-PR is the way that I'm using codec-like objects to encapsulate the parameterization of the model, separately from the model definition itself and separately from the data and the parameter values themselves.  One of the parameterizations is required to reproduce the SQUAREM results (the accelerated EM convergence properties depend on the parameterization) and the other parameterization is for helpful for unconstrained black-box minimization of the negative log likelihood.  I wonder if there is a more standard way to do this kind of thing?

**Comment by josef-pkt** [comment]
@argriffing I don't really understand your first part (and didn't really try).
I didn't know what 'aitchison' is. Based on the mailing list link, we use something similar when we use the Multinomial Logit transform to restrict values to be in the unit simplex.
about the codec-like object: The only thing I found strange was the `codec` language. We use in statsmodels parameter `transformation`, that can be turned on or off, quite often in optimization, and Kerby has one model where he outsourced different parameterizations in a separate class similar to your `codec` to keep all the transformation/reparameterization code in one location.

**Comment by argriffing** [comment]
@josef-pkt I'm not completely certain what you mean by the Multinomial Logit transform.  If you have access to "The Isometric Log-Ratio Transform for Probabilistic Multi-Label Anatomical Shape Representation" then is this the same as the ALR aka LogOdds transform defined by equations (1) -- (4) in that paper?  This seems plausible if you are referring to http://statsmodels.sourceforge.net/devel/_modules/statsmodels/genmod/generalized_estimating_equations.html in statsmodels where I found "multinomial logit transform" mentioned.  According to that paper, the ILR transform is like a better variant of ALR because ILR is more symmetric between its components.  In other words, ILR does not have a 'distinguished category' so it is less prone to certain weirdnesses or biases.

Regarding the strange `codec` language, I might change it to `transformation` or `parameterization`, but those words are longer :)  Also I like the way that the `codec` concept neatly aggregates the forward and inverse transformations and has the connotation that the 'encoded' form is less human-readable than the 'decoded' form.  But maybe it's not worth using the 'codec' notation if it is too non-standard.

**Comment by josef-pkt** [comment]
@argriffing ALR^{-1} equation (3) is the Multinomial Logit. GEE has it as a separate link/transformation function, but it's built into the standard MNLogit models like the one in statsmodels.discrete.
I didn't try to read and understand the ILR in the paper, from a quick look it's not simple.
For statistics one important point of the parameterization is also that it is easy to get various derivatives and to have a useful interpretation. I don't know if ILR would qualify.
(Also I didn't look yet in details at general models for compositional data, where I might bump into this again.)

My problem with codecs for text is that most of the time I don't remember if I'm `encoding` or `decoding`, what's the `transform` and what's the `transform.inverse`.

**Comment by argriffing** [comment]
Sorry this PR has stalled. I don't think it's blocked by anything insurmountable, but here's a collection of minor things that have slowed down my review attempt. I'm just noting these before I forget about them completely.
1) The SQUAREM reference "Simple and Globally Convergent Methods for Accelerating the Convergence of Any EM Algorithm" RAVI VARADHAN and CHRISTOPHE ROLAND 2008 somewhat depends on the earlier 2004 paper "Squared Extrapolation Methods (SQUAREM): A New Class of Simple and Efficient Numerical Schemes for Accelerating the Convergence of the EM Algorithm" by the same authors. Right now I don't remember exactly the nature of this dependence, but I remember having been confused about some things in the 2008 paper which were clarified when I found the 2004 paper. The 2004 paper is long, like 70 pages.
2) This PR uses the terminology 'merit function', but grepping 'merit' in the ~100 pages of the SQUAREM literature gives no hits.
3) Although EM can be seen as a fixed-point problem, it has additional structure.
4) The SQUAREM paper gives some results, but I seem to remember that they were not sufficient for testing the correctness of the implementation in this PR, at least not for all of the implemented features.
5) The published results use relatively complicated models.
6) There are multiple SQUAREM variants.

**Comment by kyleabeauchamp** [comment]
FWIW, I just implemented a related algorithm (http://link.springer.com/article/10.1007/s11222-009-9166-3) and found it to be quite performant and easy to implement.  That paper too has some gaps to fill, such as using the the pseudo-inverse instead of the inverse, the correct way to initialize the first handful of steps, whether or not to reject non-descent moves, etc.  However, I already found my naive implementation to be more stable than the accelerated method currently in scipy.  See gist:

https://gist.github.com/kyleabeauchamp/44922a7d458b99b38e7d7bd1a66e22d0#file-fixed_point_zhou-py

**Comment by pv** [comment]
That looks a bit like the anderson method, iow., 
`scipy.optimize.anderson(lambda x: F(x) - x, x0, w0=0, alpha=1, M=p)`

**Comment by kyleabeauchamp** [comment]
Good catch, it does seem nearly the same except without a line search.

**Comment by mdhaber** [comment]
Looks like development has been on hold for a while. Since there are merge conflicts and [several points to address](https://github.com/scipy/scipy/pull/4460#issuecomment-95967089), I'll go ahead and close, but all should feel free to re-open if they'd like to finish it up.

**Pull request closed without merging**